client.go

```diff
@@ -0,0 +1,289 @@
+package cfdns
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"mime"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/simplesurance/cfdns/logs"
+	"github.com/simplesurance/cfdns/retry"
+)
+
+const (
+	retryFirsDelay   = 2 * time.Second
+	retryMaxDelay    = 30 * time.Second
+	retryFactor      = 2
+	retryMaxAttempts = 6
+	itemsPerPage     = 500
+)
+
+func NewClient(creds Credentials, options ...Option) *Client {
+	ret := Client{
+		cfg:   applyOptions(options...),
+		creds: creds,
+	}
+
+	return &ret
+}
+
+type Client struct {
+	cfg   *settings
+	creds Credentials
+}
+
+func runWithRetry[TREQ any, TRESP commonResponseSetter](
+	ctx context.Context,
+	logger *logs.Logger,
+	req *request[TREQ],
+) (
+	resp *response[TRESP],
+	_ error,
+) {
+	reterr := retry.ExpBackoff(ctx, logger, retryFirsDelay, retryMaxDelay,
+		retryFactor, retryMaxAttempts, func() error {
+			var err error
+			resp, err = runOnce[TREQ, TRESP](ctx, logger, req)
+			return err
+		})
+
+	return resp, reterr
+}
+
+// runOnce sends an HTTP request, parses and returns the response.
+// Permanent errors are wrapped with retry.PermanentError. Any error returned
+// from the server is wrapped with HTTPError. If the error is a valid
+// CloudFlare error, it is also wrapped with CloudFlareError.
+func runOnce[TREQ any, TRESP commonResponseSetter](
```

Comment:
Restriction of the go language. I can declare a type parameter on a function or on a struct, but not on a method.

This is valid:

```go
func f[T any](p T) {}
```

This is not:

```go
type str struct{}

func(s str) [T any]f(p T) {} // method must have no type parameters (syntax)
```

I regret storing the client on `Request`. I think I'll move it to a parameter, so it will become:

```go
func sendRequest[TREQ any, TRESP commonResponseSetter](
	ctx context.Context,
	client *Client,
	logger *logs.Logger,
	treq *request[TREQ],
) (
	tresp *response[TRESP],
	err error,
) {
}
```

WDYT?

Reply:
Yeah, I moved client to its own parameter, instead of being a member of `request`.